src/main/java/org/jenkinsci/plugins/workflow/libs/SCMSourceRetriever.java

```diff
@@ -293,4 +301,62 @@ public FormValidation doCheckLibraryPath(@QueryParameter String libraryPath) {
 
     }
 
+    @Restricted(DoNotUse.class)
+    @Extension
+    public static class WorkspaceListener extends ItemListener {
+
+        @Override
+        public void onDeleted(Item item) {
+            deleteLibsDir(item, item.getFullName());
+        }
+
+        @Override
+        public void onLocationChanged(Item item, String oldFullName, String newFullName) {
+            deleteLibsDir(item, oldFullName);
+        }
+
+        private static void deleteLibsDir(Item item, String itemFullName) {
+            if (item instanceof Job
+                    && item.getClass()
+                            .getName()
+                            .equals("org.jenkinsci.plugins.workflow.job.WorkflowJob")) {
+                synchronized (item) {
+                    String base =
+                            expandVariablesForDirectory(
+                                    Jenkins.get().getRawWorkspaceDir(),
+                                    itemFullName,
+                                    item.getRootDir().getPath());
+                    FilePath baseWorkspace = new FilePath(new File(base));
+                    if (baseWorkspace != null) {
+                        FilePath dir = baseWorkspace.withSuffix(getFilePathSuffix() + "libs");
+                        try {
+                            if (dir.isDirectory()) {
+                                LOGGER.log(
+                                        Level.INFO,
+                                        () -> "Deleting obsolete shared library workspace " + dir);
+                                dir.deleteRecursive();
+                            }
+                        } catch (IOException | InterruptedException e) {
+                            LOGGER.log(
+                                    Level.WARNING,
+                                    e,
+                                    () -> "Could not delete obsolete shared library workspace " + dir);
+                        }
+                    }
+                }
+            }
+        }
+
+        private static String expandVariablesForDirectory(
```

Comment:
I guess we need to do this because if the item is moved, it will be too late to look up its original workspace location by the time we get the notification?

Reply:
Yes, exactly.

Reply:
Took me at least a while to get there…maybe deserves a comment, linking to https://github.com/jenkinsci/jenkins/blob/f03183ab09ce5fb8f9f4cc9ccee42a3c3e6b2d3e/core/src/main/java/jenkins/model/Jenkins.java#L2567-L2576? (And maybe we should just remove the API restriction there so that at some point this code could be replaced with a call to the official method?)

Reply:
I updated the comment. I feel that there are enough bits and pieces of core duplicate in Pipeline that this isn't materially changing the situation. That and the… "overhead" of making API changes to core makes me want to leave this alone.